feat: Add a csv report of xblocks used in courses (redwood)

samuelallan72 opened the pull request:
## Description

Add a management command for generating a csv report of xblocks used in courses.

Usage (with common management command help removed for clarity):

```
$ python manage.py lms xblock_list_csv --help
...
usage: manage.py xblock_list_csv [-h] [--exclude-core-xblocks] [--version] [-v {0,1,2,3}] [--settings SETTINGS] [--pythonpath PYTHONPATH] [--traceback] [--no-color]
                                 [--force-color] [--skip-checks]
                                 filename

Generate a CSV file with all published components used in courses with their xblock type.

positional arguments:
  filename              Path to the CSV output file. Use '-' to print to stdout.

options:
  -h, --help            show this help message and exit
  --exclude-core-xblocks
                        Exclude components using core XBlocks: (html, problem, video). Default: false
...
```

### Examples:

Print complete csv to stdout:

```
python manage.py lms xblock_list_csv - 
```


Save complete csv to components.csv:

```
python manage.py lms xblock_list_csv components.csv
```


Save csv excluding components using core xblocks to components.csv:

```
python manage.py lms xblock_list_csv --exclude-core-xblocks components.csv
```

Example csv output files:

[components-excluded.csv](https://github.com/user-attachments/files/23891526/components-excluded.csv)
[components.csv](https://github.com/user-attachments/files/23891527/components.csv)
[components-stdout.csv](https://github.com/user-attachments/files/23891528/components-stdout.csv)


## Supporting information

Private-ref: https://tasks.opencraft.com/browse/BB-10225

## Testing instructions

- deploy this branch to a redwood devstack
- ensure there is a range of published course content across at least two courses on the devstack
- start an lms shell: `tutor dev exec lms -- bash`
- run `python manage.py lms xblock_list_csv -`
- verify the csv data is printed to the console
- run `python manage.py lms xblock_list_csv components.csv`
- verify the csv data is saved to components.csv
- run `python manage.py lms xblock_list_csv - > components-stdout.csv`
- run `diff components.csv components-stdout.csv` and verify the diff is empty (to verify printing to stdout has the same report as writing to file)
- verify components.csv contains rows as expected - check that all courses are included and the data makes sense
- run `python manage.py lms xblock_list_csv --exclude-core-xblocks components-excluded.csv`
- in `components-excluded.csv`, verify there are no rows for the core xblocks (html, problem, video)

pkulkark commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1502,6 +1502,32 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
+    try:
+        task_api.submit_xblocks_list_csv(
+            request,
+            course_key,
```
This should be the course id or the `course` variable right?

pkulkark replied:
Actually no, it should be `course_key`. In that case, do we need the step `course = get_course_by_id(course_key)`? It doesn't seem to be used for anything.

pkulkark commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1502,6 +1502,32 @@ def get_students_features(request, course_id, csv=False):  # pylint: disable=red
+        )
+        success_status = SUCCESS_MESSAGE_TEMPLATE.format(report_type=report_type)
+    except Exception as e:
+        raise self.api_error(status.HTTP_400_BAD_REQUEST, str(e), 'Requested task is already running')
```
This is a module-level function, so there's no `self.api_error` available.

samuelallan72 replied:
Ah thanks, this is leftover from porting from master.

pkulkark commented on lms/djangoapps/instructor_task/tasks_helper/components.py:
```diff
@@ -0,0 +1,88 @@
+    succeeded_count = 0
+    for overview in overviews:
+        try:
+            course = modulestore().get_course(overview.id)
```
Can't we directly use the `course_id` instead of fetching the overviews and using the `overview.id`?

samuelallan72 replied:
Yeah this is leftover code from when I was running it for all courses. I didn't refactor, because I wasn't sure if we'd use it yet.

pkulkark commented on lms/djangoapps/instructor_task/tasks_helper/components.py:
```diff
@@ -0,0 +1,88 @@
+                    "Subsection Name": subsection.display_name,
+                    "Unit Name": unit.display_name,
+                    "Component Name": component.display_name,
+                    "Xblock Type": component.location.block_type,
```
This would just return all the component types in a course right? I wonder if it would be better to filter out the default ones (like html, problem, video) from external ones (like pdf, timeline, short answer, etc)?

samuelallan72 replied:
Ah maybe, not sure. We can probably ask the client - they may be interested in all components, for stats like which percentage of the components are third party xblocks.

samuelallan72 committed:
feat: add a csv report of xblocks used in courses

This is added via an lms management command:

```
$ python manage.py lms xblock_list_csv --help
...
usage: manage.py xblock_list_csv [-h] [--exclude-core-xblocks] [--version] [-v {0,1,2,3}] [--settings SETTINGS] [--pythonpath PYTHONPATH] [--traceback] [--no-color]
                                 [--force-color] [--skip-checks]
                                 filename

Generate a CSV file with all published components used in courses with their xblock type.

positional arguments:
  filename              Path to the CSV output file. Use '-' to print to stdout.

options:
  -h, --help            show this help message and exit
  --exclude-core-xblocks
                        Exclude components using core XBlocks: (html, problem, video). Default: false
...
```

Example usage:

```
python manage.py lms xblock_list_csv --exclude-core-xblocks components.csv
```

Private-ref: https://tasks.opencraft.com/browse/BB-10225

pkulkark reviewed (approved):
LGTM 👍 

- [x] I tested this: I couldn't get my devstack running but verified the generated reports are as expected.
- [x] I read through the code
